ReactAndroid/src/main/java/com/facebook/react/modules/network/OkHttpClientProvider.java

```diff
@@ -43,18 +59,51 @@ public static OkHttpClient getOkHttpClient() {
     }
     return sClient;
   }
+
+  private static class EvictIdleConnectionsTask extends AsyncTask {
+    @Override
+    protected Object doInBackground(Object[] objects) {
+      for (OkHttpClient client: sClients) {
+        client.connectionPool().evictAll();
+      }
+      return null;
+    }
+  }
+
+  public static void addNetworkListenerToEvictClientConnectionsOnDisconnect(Context context) {
+    BroadcastReceiver br = new BroadcastReceiver() {
+      @Override
+      public void onReceive(Context context, Intent intent) {
+        Bundle extras = intent.getExtras();
+        NetworkInfo info = extras.getParcelable("networkInfo");
+        NetworkInfo.State state = info.getState();
+        if (state == NetworkInfo.State.DISCONNECTED) {
+          new EvictIdleConnectionsTask().execute();
+        }
+      }
+    };
+    final IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(ConnectivityManager.CONNECTIVITY_ACTION);
+    context.registerReceiver(br, intentFilter);
+  }
   
   // okhttp3 OkHttpClient is immutable
   // This allows app to init an OkHttpClient with custom settings.
   public static void replaceOkHttpClient(OkHttpClient client) {
     sClient = client;
+    sClients.add(client);
   }
 
   public static OkHttpClient createClient() {
+    OkHttpClient client;
     if (sFactory != null) {
-      return sFactory.createNewNetworkModuleClient();
+      client = sFactory.createNewNetworkModuleClient();
     }
-    return createClientBuilder().build();
+    else {
+      client = createClientBuilder().build();
+    }
+    sClients.add(client);
```

Comment:
My inclination would actually be to stick this logic in `createClientBuilder`, rather than here. That leaves apps with full flexibility to drop or customize this logic if they choose to set their own factory.

An extra friendly version of that approach would be to pull out this line `sClients.add(client)` into a public static method, like `enableTls12OnPreLollipop`. That way, such an app can call that method if they want to opt back into this feature of `OkHttpClientProvider`.

Reply:
We can't do this in `createClientBuilder` because that returns a client builder, not a client. However, flexibility is a good point -- I'll drop the `sClients.add` for `replaceOkHttpClient` and `createClient` with an sFactory, and put `sClients.add(client);` into a helper method.

Reply:
> We can't do this in createClientBuilder because that returns a client builder, not a client. 

Hmm, indeed. I guess I was misled by the variable and parameter names -- it gets called `client` here and in passing to `enableTls12OnPreLollipop`.

This other solution is good.